File: webagents/agents/skills/local/testrunner/parser.py
"""
Test specification parser.

Parses Markdown test files with YAML frontmatter into structured test specs.
"""

import re
import yaml
from pathlib import Path
from typing import Dict, Any, List, Optional
from dataclasses import dataclass, field


@dataclass
class TestCase:
    """A single test case."""
    name: str
    request: Optional[Dict[str, Any]] = None
    flow: Optional[List[str]] = None
    assertions: List[str] = field(default_factory=list)
    strict: Optional[Dict[str, Any]] = None
    expected: str = "success"


@dataclass
class TestSpec:
    """A parsed test specification."""
    name: str
    version: str
    transport: str
    type: str = "single-agent"
    tags: List[str] = field(default_factory=list)
    timeout: int = 60
    depends_on: List[str] = field(default_factory=list)
    description: str = ""
    setup: str = ""
    agents: List[Dict[str, Any]] = field(default_factory=list)
    cases: List[TestCase] = field(default_factory=list)


class TestParser:
    """
    Parser for Markdown compliance test files.
    
    Parses files with structure:
    - YAML frontmatter
    - # Title
    - Description paragraph
    - ## Setup
    - ## Test Cases
    - ### N. Case Name
    """
    
    def parse(self, path: Path) -> Dict[str, Any]:
        """
        Parse a test file into a structured spec.
        
        Args:
            path: Path to the markdown file
            
        Returns:
            Parsed test specification as dict
        """
        content = path.read_text()
        
        # Extract frontmatter
        frontmatter, body = self._extract_frontmatter(content)
        
        # Parse body sections
        title, description = self._extract_title_description(body)
        setup = self._extract_section(body, "Setup")
        agents = self._parse_agents(setup) if setup else []
        cases = self._parse_test_cases(body)
        
        spec = TestSpec(
            name=frontmatter.get("name", path.stem),
            version=frontmatter.get("version", "1.0"),
            transport=frontmatter.get("transport", "completions"),
            type=frontmatter.get("type", "single-agent"),
            tags=frontmatter.get("tags", []),
            timeout=frontmatter.get("timeout", 60),
            depends_on=frontmatter.get("depends_on", []),
            description=description,
            setup=setup or "",
            agents=agents,
            cases=cases,
        )
        
        return self._spec_to_dict(spec)
    
    def _extract_frontmatter(self, content: str) -> tuple[Dict[str, Any], str]:
        """Extract YAML frontmatter from markdown."""
        match = re.match(r'^---\n(.*?)\n---\n(.*)$', content, re.DOTALL)
        if match:
            try:
                frontmatter = yaml.safe_load(match.group(1)) or {}
            except yaml.YAMLError:
                frontmatter = {}
            body = match.group(2)
        else:
            frontmatter = {}
            body = content
        return frontmatter, body
    
    def _extract_title_description(self, body: str) -> tuple[str, str]:
        """Extract title and description from body."""
        # Find H1 title
        title_match = re.search(r'^# (.+)$', body, re.MULTILINE)
        title = title_match.group(1) if title_match else ""
        
        # Description is text between title and first ## section
        if title_match:
            after_title = body[title_match.end():]
            desc_match = re.match(r'\n+(.+?)(?=\n## |\Z)', after_title, re.DOTALL)
            description = desc_match.group(1).strip() if desc_match else ""
        else:
            description = ""
        
        return title, description
    
    def _extract_section(self, body: str, section_name: str) -> Optional[str]:
        """Extract content of a ## section."""
        pattern = rf'^## {section_name}\s*\n(.*?)(?=\n## |\Z)'
        match = re.search(pattern, body, re.MULTILINE | re.DOTALL)
        return match.group(1).strip() if match else None
    
    def _parse_agents(self, setup: str) -> List[Dict[str, Any]]:
        """Parse agent definitions from setup section."""
        agents = []
        
        # Find ### Agent: name sections
        agent_pattern = r'### Agent: (\w+)\s*\n(.*?)(?=\n### |\Z)'
        for match in re.finditer(agent_pattern, setup, re.DOTALL):
            agent_id = match.group(1)
            agent_content = match.group(2)
            
            agent = {"id": agent_id}
            
            # Parse bullet points
            for line in agent_content.split('\n'):
                line = line.strip()
                if line.startswith('- Name:'):
                    agent["name"] = line.split(':', 1)[1].strip().strip('`')
                elif line.startswith('- Instructions:'):
                    agent["instructions"] = line.split(':', 1)[1].strip().strip('"')
                elif line.startswith('- Handoffs:'):
                    handoffs_str = line.split(':', 1)[1].strip()
                    agent["handoffs"] = [h.strip() for h in handoffs_str.strip('[]').split(',')]
                elif line.startswith('- Tools:'):
                    tools_str = line.split(':', 1)[1].strip()
                    agent["tools"] = [t.strip() for t in tools_str.strip('[]').split(',')]
            
            agents.append(agent)
        
        return agents
    
    def _parse_test_cases(self, body: str) -> List[TestCase]:
        """Parse test cases from body."""
        cases = []
        
        # Find ## Test Cases section
        cases_section = self._extract_section(body, "Test Cases")
        if not cases_section:
            return cases
        
        # Find ### N. Case Name sections
        case_pattern = r'### \d+\. (.+?)\s*\n(.*?)(?=\n### |\Z)'
        for match in re.finditer(case_pattern, cases_section, re.DOTALL):
            case_name = match.group(1).strip()
            case_content = match.group(2)
            
            case = TestCase(name=case_name)
            
            # Parse Request section
            request = self._parse_request(case_content)
            if request:
                case.request = request
            
            # Parse Flow section (for multi-agent)
            flow = self._parse_flow(case_content)
            if flow:
                case.flow = flow
            
            # Parse Assertions
            case.assertions = self._parse_assertions(case_content)
            
            # Parse Strict block
            strict = self._parse_strict(case_content)
            if strict:
                case.strict = strict
            
            # Parse Expected
            expected_match = re.search(r'\*\*Expected:\*\*\s*(\w+)', case_content)
            if expected_match:
                case.expected = expected_match.group(1)
            
            cases.append(case)
        
        return cases
    
    def _parse_request(self, content: str) -> Optional[Dict[str, Any]]:
        """Parse request section from test case."""
        match = re.search(
            r'\*\*Request:\*\*\s*\n(.*?)(?=\n\*\*|\Z)',
            content,
            re.DOTALL
        )
        if not match:
            return None
        
        request_content = match.group(1)
        
        # Extract method and path
        method_match = re.search(r'(GET|POST|PUT|DELETE|PATCH)\s+`?([^`\s]+)`?', request_content)
        if not method_match:
            return None
        
        request = {
            "method": method_match.group(1),
            "path": method_match.group(2),
        }
        
        # Extract JSON body
        json_match = re.search(r'```json\s*\n(.*?)\n```', request_content, re.DOTALL)
        if json_match:
            try:
                request["body"] = yaml.safe_load(json_match.group(1))
            except:
                request["body_raw"] = json_match.group(1)
        
        # Check for streaming
        if '(streaming)' in request_content.lower() or '"stream": true' in request_content:
            request["stream"] = True
        
        return request
    
    def _parse_flow(self, content: str) -> Optional[List[str]]:
        """Parse flow section for multi-agent tests."""
        match = re.search(r'\*\*Flow:\*\*\s*\n(.*?)(?=\n\*\*|\Z)', content, re.DOTALL)
        if not match:
            return None
        
        flow = []
        for line in match.group(1).split('\n'):
            line = line.strip()
            if re.match(r'^\d+\.', line):
                flow.append(re.sub(r'^\d+\.\s*', '', line))
        
        return flow if flow else None
    
    def _parse_assertions(self, content: str) -> List[str]:
        """Parse natural language assertions."""
        match = re.search(r'\*\*Assertions:\*\*\s*\n(.*?)(?=\n\*\*|\Z)', content, re.DOTALL)
        if not match:
            return []
        
        assertions = []
        for line in match.group(1).split('\n'):
            line = line.strip()
            if line.startswith('- '):
                assertions.append(line[2:])
        
        return assertions
    
    def _parse_strict(self, content: str) -> Optional[Dict[str, Any]]:
        """Parse strict assertions YAML block."""
        match = re.search(r'\*\*Strict:\*\*\s*\n```yaml\s*\n(.*?)\n```', content, re.DOTALL)
        if not match:
            return None
        
        try:
            return yaml.safe_load(match.group(1))
        except yaml.YAMLError:
            return None
    
    def _spec_to_dict(self, spec: TestSpec) -> Dict[str, Any]:
        """Convert TestSpec to dict."""
        return {
            "name": spec.name,
            "version": spec.version,
            "transport": spec.transport,
            "type": spec.type,
            "tags": spec.tags,
            "timeout": spec.timeout,
            "depends_on": spec.depends_on,
            "description": spec.description,
            "setup": spec.setup,
            "agents": spec.agents,
            "cases": [
                {
                    "name": case.name,
                    "request": case.request,
                    "flow": case.flow,
                    "assertions": case.assertions,
                    "strict": case.strict,
                    "expected": case.expected,
                }
                for case in spec.cases
            ],
        }

File: typescript/src/skills/testrunner/parser.ts
/**
 * Test specification parser
 * 
 * Parses Markdown test files with YAML frontmatter into structured test specs.
 */

import * as yaml from 'yaml';
import type { TestSpec, TestCase, HttpRequestConfig, AgentDefinition } from './types';

/**
 * Parser for Markdown compliance test files.
 * 
 * Parses files with structure:
 * - YAML frontmatter
 * - # Title
 * - Description paragraph
 * - ## Setup
 * - ## Test Cases
 * - ### N. Case Name
 */
export class TestParser {
  /**
   * Parse a test file content into a structured spec.
   */
  parse(content: string, filename: string = 'test'): TestSpec {
    // Extract frontmatter
    const { frontmatter, body } = this.extractFrontmatter(content);

    // Parse body sections
    const { description } = this.extractTitleDescription(body);
    const setup = this.extractSection(body, 'Setup');
    const agents = setup ? this.parseAgents(setup) : [];
    const cases = this.parseTestCases(body);

    return {
      name: (frontmatter.name as string) || this.getBasename(filename),
      version: (frontmatter.version as string) || '1.0',
      transport: (frontmatter.transport as string) || 'completions',
      type: (frontmatter.type as 'single-agent' | 'multi-agent') || 'single-agent',
      tags: (frontmatter.tags as string[]) || [],
      timeout: (frontmatter.timeout as number) || 60,
      dependsOn: (frontmatter.depends_on as string[]) || [],
      description,
      setup: setup || '',
      agents,
      cases,
    };
  }

  /**
   * Extract YAML frontmatter from markdown.
   */
  private extractFrontmatter(content: string): { frontmatter: Record<string, unknown>; body: string } {
    const match = content.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
    
    if (match) {
      try {
        const frontmatter = yaml.parse(match[1]) || {};
        return { frontmatter, body: match[2] };
      } catch {
        return { frontmatter: {}, body: match[2] };
      }
    }
    
    return { frontmatter: {}, body: content };
  }

  /**
   * Extract title and description from body.
   */
  private extractTitleDescription(body: string): { title: string; description: string } {
    // Find H1 title
    const titleMatch = body.match(/^# (.+)$/m);
    const title = titleMatch ? titleMatch[1] : '';

    // Description is text between title and first ## section
    let description = '';
    if (titleMatch) {
      const afterTitle = body.slice(titleMatch.index! + titleMatch[0].length);
      const descMatch = afterTitle.match(/^\n+(.+?)(?=\n## |\n*$)/s);
      description = descMatch ? descMatch[1].trim() : '';
    }

    return { title, description };
  }

  /**
   * Extract content of a ## section.
   */
  private extractSection(body: string, sectionName: string): string | null {
    const pattern = new RegExp(`^## ${sectionName}\\s*\\n([\\s\\S]*?)(?=\\n## |$)`, 'm');
    const match = body.match(pattern);
    return match ? match[1].trim() : null;
  }

  /**
   * Parse agent definitions from setup section.
   */
  private parseAgents(setup: string): AgentDefinition[] {
    const agents: AgentDefinition[] = [];

    // Find ### Agent: name sections
    const agentPattern = /### Agent: (\w+)\s*\n([\s\S]*?)(?=\n### |$)/g;
    let match;

    while ((match = agentPattern.exec(setup)) !== null) {
      const agentId = match[1];
      const agentContent = match[2];

      const agent: AgentDefinition = { id: agentId };

      // Parse bullet points
      for (const line of agentContent.split('\n')) {
        const trimmed = line.trim();
        if (trimmed.startsWith('- Name:')) {
          agent.name = trimmed.split(':', 2)[1].trim().replace(/`/g, '');
        } else if (trimmed.startsWith('- Instructions:')) {
          agent.instructions = trimmed.split(':', 2)[1].trim().replace(/"/g, '');
        } else if (trimmed.startsWith('- Handoffs:')) {
          const handoffsStr = trimmed.split(':', 2)[1].trim();
          agent.handoffs = handoffsStr
            .replace(/[\[\]]/g, '')
            .split(',')
            .map(h => h.trim())
            .filter(Boolean);
        } else if (trimmed.startsWith('- Tools:')) {
          const toolsStr = trimmed.split(':', 2)[1].trim();
          agent.tools = toolsStr
            .replace(/[\[\]]/g, '')
            .split(',')
            .map(t => t.trim())
            .filter(Boolean);
        }
      }

      agents.push(agent);
    }

    return agents;
  }

  /**
   * Parse test cases from body.
   */
  private parseTestCases(body: string): TestCase[] {
    const cases: TestCase[] = [];

    // Find ## Test Cases section
    const casesSection = this.extractSection(body, 'Test Cases');
    if (!casesSection) {
      return cases;
    }

    // Find ### N. Case Name sections
    const casePattern = /### \d+\. (.+?)\s*\n([\s\S]*?)(?=\n### |$)/g;
    let match;

    while ((match = casePattern.exec(casesSection)) !== null) {
      const caseName = match[1].trim();
      const caseContent = match[2];

      const testCase: TestCase = {
        name: caseName,
        assertions: [],
        expected: 'success',
      };

      // Parse Request section
      const request = this.parseRequest(caseContent);
      if (request) {
        testCase.request = request;
      }

      // Parse Flow section (for multi-agent)
      const flow = this.parseFlow(caseContent);
      if (flow) {
        testCase.flow = flow;
      }

      // Parse Assertions
      testCase.assertions = this.parseAssertions(caseContent);

      // Parse Strict block
      const strict = this.parseStrict(caseContent);
      if (strict) {
        testCase.strict = strict;
      }

      // Parse Expected
      const expectedMatch = caseContent.match(/\*\*Expected:\*\*\s*(\w+)/);
      if (expectedMatch) {
        testCase.expected = expectedMatch[1] as 'success' | 'failure';
      }

      cases.push(testCase);
    }

    return cases;
  }

  /**
   * Parse request section from test case.
   */
  private parseRequest(content: string): HttpRequestConfig | null {
    const match = content.match(/\*\*Request:\*\*\s*\n([\s\S]*?)(?=\n\*\*|$)/);
    if (!match) {
      return null;
    }

    const requestContent = match[1];

    // Extract method and path
    const methodMatch = requestContent.match(/(GET|POST|PUT|DELETE|PATCH)\s+`?([^`\s]+)`?/);
    if (!methodMatch) {
      return null;
    }

    const request: HttpRequestConfig = {
      method: methodMatch[1],
      path: methodMatch[2],
    };

    // Extract JSON body
    const jsonMatch = requestContent.match(/```json\s*\n([\s\S]*?)\n```/);
    if (jsonMatch) {
      try {
        request.body = yaml.parse(jsonMatch[1]);
      } catch {
        // Body parsing failed, skip
      }
    }

    // Check for streaming
    if (
      requestContent.toLowerCase().includes('(streaming)') ||
      requestContent.includes('"stream": true')
    ) {
      request.stream = true;
    }

    return request;
  }

  /**
   * Parse flow section for multi-agent tests.
   */
  private parseFlow(content: string): string[] | null {
    const match = content.match(/\*\*Flow:\*\*\s*\n([\s\S]*?)(?=\n\*\*|$)/);
    if (!match) {
      return null;
    }

    const flow: string[] = [];
    for (const line of match[1].split('\n')) {
      const trimmed = line.trim();
      if (/^\d+\./.test(trimmed)) {
        flow.push(trimmed.replace(/^\d+\.\s*/, ''));
      }
    }

    return flow.length > 0 ? flow : null;
  }

  /**
   * Parse natural language assertions.
   */
  private parseAssertions(content: string): string[] {
    const match = content.match(/\*\*Assertions:\*\*\s*\n([\s\S]*?)(?=\n\*\*|$)/);
    if (!match) {
      return [];
    }

    const assertions: string[] = [];
    for (const line of match[1].split('\n')) {
      const trimmed = line.trim();
      if (trimmed.startsWith('- ')) {
        assertions.push(trimmed.slice(2));
      }
    }

    return assertions;
  }

  /**
   * Parse strict assertions YAML block.
   */
  private parseStrict(content: string): Record<string, unknown> | null {
    const match = content.match(/\*\*Strict:\*\*\s*\n```yaml\s*\n([\s\S]*?)\n```/);
    if (!match) {
      return null;
    }

    try {
      return yaml.parse(match[1]);
    } catch {
      return null;
    }
  }

  /**
   * Get basename from filename.
   */
  private getBasename(filename: string): string {
    const parts = filename.split('/');
    const name = parts[parts.length - 1];
    return name.replace(/\.[^.]+$/, '');
  }
}
